src/super_context.ml

```diff
@@ -865,53 +865,120 @@ module PP = struct
           mli) in
     { m with impl ; intf }
 
+  (* laziness is needed to support the case where a user is preprocessing a
+     reason file with action but linting with ppx. *)
+  let lint_module sctx ~(source : Module.t) ~(ast : Module.t Lazy.t) ~dir
+        ~dep_kind ~lint ~lib_name ~scope =
+    let alias = Alias.lint ~dir in
+    match Preprocess_map.find source.name lint with
+    | No_preprocessing -> ()
+    | Action action ->
+      let action = Action.U.Chdir (root_var, action) in
+      Module.iter source ~f:(fun _ (src : Module.File.t) ->
+        let digest_path =
+          Alias.add_action_dep
+            ~action:(Some action)
+            ~action_deps:[Dep_conf.File (String_with_vars.virt __POS__ src.name)]
+            (aliases sctx) alias in
+        let src = Path.relative dir src.name in
+        add_rule sctx
+          (Build.path src
+           >>^ (fun _ -> [src])
+           >>>
+           Build.progn
+             [ Action.run sctx
+                 action
+                 ~dir
+                 ~dep_kind:Required
+                 ~targets:(Static [])
+                 ~scope
+             ; Build.create_file digest_path
+             ])
+      )
+    | Pps { pps; flags } ->
+      let ppx_exe = get_ppx_driver sctx pps ~dir ~dep_kind in
+      Module.iter (Lazy.force ast) ~f:(fun kind src ->
+        let src_path = Path.relative dir src.name in
+        let args =
+          [ Arg_spec.As flags
+          ; As (cookie_library_name lib_name)
+          ; Ml_kind.ppx_driver_flag kind
+          ; Dep src_path
+            (* ; A "--null" *)
+          ] in
+        let digest_path =
+          Alias.add_action_dep (aliases sctx) alias
+            ~action_deps:[Dep_conf.File (String_with_vars.virt __POS__ src.name)]
+            ~action:(
+              let (args, paths) = Arg_spec.expand ~dir args () in
+              Some (Action.U.Run (
+                String_with_vars.virt __POS__ (Path.to_string ppx_exe),
+                List.map ~f:(String_with_vars.virt __POS__)
+                  (args @ (List.map ~f:Path.to_string (Pset.elements paths)))
+              ))
+            ) in
+        add_rule sctx
+          (Build.progn
+             [ Build.run ~context:sctx.context (Ok ppx_exe) args
+             ; Build.create_file digest_path
+             ])
+      )
+
   (* Generate rules to build the .pp files and return a new module map where all filenames
      point to the .pp files *)
-  let pped_modules sctx ~dir ~dep_kind ~modules ~preprocess ~preprocessor_deps ~lib_name
-        ~scope =
+  let pped_linted_modules sctx ~dir ~dep_kind ~modules ~lint ~preprocess
+        ~preprocessor_deps ~lib_name ~scope =
```

Comment:
Nitpicking: `pped_modules` was supposed to be "pre-processed modules". But linted modules doesn't make much sense to me given that the linting steps produces nothing. What about: `preprocess_and_lint_modules`?

Reply:
OK sure. Let's keep it shorter with `pp_and_lint_modules`. it's called pp everywhere else anyway.